# Análise do risco de inadimplência dos mutuários

Meu projeto foi preparar um relatório para a divisão de empréstimos de um banco. onde precisei descobrir se o estado civil de um cliente e o número de filhos têm impacto sobre se ele deixará de pagar um empréstimo. O banco já tem alguns dados sobre a capacidade de crédito dos clientes.

Meu relatório tem como objetivo  criar uma **pontuação de crédito** de um cliente em potencial. A **contagem de crédito** é usada para avaliar a capacidade de um devedor em potencial de pagar seu empréstimo.


##  Arquivo de dados e as informações gerais.


In [1]:
# Carregando todas as bibliotecas
import pandas as pd

# Carregue os dados
dados =pd.read_csv('/datasets/credit_scoring_eng.csv')


In [2]:
# Exibindo as primeiras 10 linhas
dados.head(10)  

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

##  Exploração de dados

**Descrição dos dados**
- `children` - o número de crianças na família
- `days_employed` - experiência de trabalho em dias
- `dob_years` - idade do cliente em anos
- `education` - educação do cliente
- `education_id` - identificador de educação
- `family_status` - estado civil do cliente
- `family_status_id` - identificador de estado civil
- `gender` - gênero do cliente
- `income_type` - tipo de emprego
- `debt` - havia alguma dívida no pagamento do empréstimo
- `total_income` - renda mensal
- `purpose` - o objetivo de obter um empréstimo


In [3]:
# Vamos ver quantas linhas e colunas nosso conjunto de dados tem

print(f' O número de linhas é {dados.shape[0]} e o número de colunas é {dados.shape[1]}.')

 O número de linhas é 21525 e o número de colunas é 12.


In [4]:
# vamos exibir as primeiras 20 linhas
dados.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  Secondary Education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  SECONDARY EDUCATION             1   
8          2   -6929.865299         35    BACHELOR'S DEGREE             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0   -1846.641941         54         some college             2   
14         0   -1844.956182         56    bachelor's degree             0   
15         1    -972.364419         26  secondary education             1   
16         0   -1719.934226         35  secondary education             1   
17         0   -2369.999720         33    bachelor's degree             0   
18         0  400281.136913         53  secondary education             1   
19         0  -10038.818549         48  SECONDARY EDUCATION             1   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   
15            married                 0      F    employee     0   
16            married                 0      F    employee     0   
17  civil partnership                 1      M    employee     0   
18    widow / widower                 2      F     retiree     0   
19           divorced                 3      F    employee     0   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate  
15     18691.345         construction of own property  
16   


- Na coluna days_employed o qual seria os dias trabalhados , existem numeros negativos alem de valores ausentes .
- É possivel observar valores ausentes  na coluna total_income - renda mensal, podendo causar problemas para análise.
- Existe uma diferença de grafia dos dados na coluna "education" que podemos investigar.



In [5]:
# Obter informações sobre dados
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


- Poderia eliminar as colunas com numeros ausentes atraves do metodo dropna() e depois aplicar o metodo astype(int) porém não teriamos os numeros ausentes para fazer a comparação , e não é possivel aplicar o metodo astype devido aos numeros ausentes são strings , então de fato  poderia trocar atraves do metodo replace por suas medianas ou apenas aplicar o metodo astype depois de retirar os numeros ausentes 

In [18]:
#metodo para deixar os valores da coluna de dias trabalhados inteiro
dados_day_employed_sem_ausentes = dados['days_employed'].dropna(axis=0)
print(dados_day_employed_sem_ausentes.head(15))
dados_day_employed_sem_ausentes.astype(int)
#apenas troquei a variavel devido ao fato se eu eliminasse os numeros ausentes iriam estragar minhas análises abaixo

0      -8437.673028
1      -4024.803754
2      -5623.422610
3      -4124.747207
4     340266.072047
5       -926.185831
6      -2879.202052
7       -152.779569
8      -6929.865299
9      -2188.756445
10     -4171.483647
11      -792.701887
13     -1846.641941
14     -1844.956182
15      -972.364419
Name: days_employed, dtype: float64


0         -8437
1         -4024
2         -5623
3         -4124
4        340266
          ...  
21520     -4529
21521    343937
21522     -2113
21523     -3112
21524     -1984
Name: days_employed, Length: 19351, dtype: int64

In [6]:
print(dados.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None



- Existem valores ausentes nas colunas days_employed e  total_income  
- a coluna de dias trabalhados foi mudada para int através do método de inteiro e do metodo de arredondamento

In [7]:
# Vejamos a tabela filtrada com valores ausentes na primeira coluna com dados ausentes

days_employed_null = dados['days_employed'].isnull()
dados[days_employed_null].head(15)


children  days_employed  dob_years            education  education_id  \
12          0            NaN         65  secondary education             1   
26          0            NaN         41  secondary education             1   
29          0            NaN         63  secondary education             1   
41          0            NaN         50  secondary education             1   
55          0            NaN         54  secondary education             1   
65          0            NaN         21  secondary education             1   
67          0            NaN         52    bachelor's degree             0   
72          1            NaN         32    bachelor's degree             0   
82          2            NaN         50    bachelor's degree             0   
83          0            NaN         52  secondary education             1   
90          2            NaN         35    bachelor's degree             0   
94          1            NaN         34    bachelor's degree             0   
96          0            NaN         44  SECONDARY EDUCATION             1   
97          0            NaN         47    bachelor's degree             0   
120         0            NaN         46  secondary education             1   

         family_status  family_status_id gender    income_type  debt  \
12   civil partnership                 1      M        retiree     0   
26             married                 0      M  civil servant     0   
29           unmarried                 4      F        retiree     0   
41             married                 0      F  civil servant     0   
55   civil partnership                 1      F        retiree     1   
65           unmarried                 4      M       business     0   
67             married                 0      F        retiree     0   
72             married                 0      M  civil servant     0   
82             married                 0      F       employee     0   
83             married                 0      M       employee     0   
90             married                 0      F       employee     0   
94   civil partnership                 1      F       business     0   
96             married                 0      F       employee     0   
97             married                 0      F       employee     0   
120            married                 0      F       employee     0   

     total_income                                   purpose  
12            NaN                         to have a wedding  
26            NaN                                 education  
29            NaN                    building a real estate  
41            NaN                  second-hand car purchase  
55            NaN                         to have a wedding  
65            NaN  transactions with commercial real estate  
67            NaN       purchase of the house for my family  
72            NaN  transactions with commercial real estate  
82            NaN                                   housing  
83            NaN                                   housing  
90            NaN                      housing transactions  
94            NaN                          having a wedding  
96            NaN               buy residential real estate  
97            NaN                         profile education  
120           NaN                      university education

- Observamos que em days_employed todos os valores ausentes são simétricos com os valores ausentes de total_income, porém não podemos ter certeza dessa suposição ,assim precisamos de mais testes parar afirmar tal suposição.


In [8]:
# Verificando a quantidade de valores ausentes na base de dados
dados.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [9]:
# Verificando as informações dos dados filtrados e se as colunas "days_employed" e "total_income" possuem a mesma quantidade de valores não nulos
dados[days_employed_null].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2174 entries, 12 to 21510
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          2174 non-null   int64  
 1   days_employed     0 non-null      float64
 2   dob_years         2174 non-null   int64  
 3   education         2174 non-null   object 
 4   education_id      2174 non-null   int64  
 5   family_status     2174 non-null   object 
 6   family_status_id  2174 non-null   int64  
 7   gender            2174 non-null   object 
 8   income_type       2174 non-null   object 
 9   debt              2174 non-null   int64  
 10  total_income      0 non-null      float64
 11  purpose           2174 non-null   object 
dtypes: float64(2), int64(5), object(5)
memory usage: 220.8+ KB


In [10]:
#Verificando os valores ausentes
dados.loc[(dados["total_income" ].isnull())]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

**Conclusão intermediária**

- Para cada valor "days_employed" ausente, o valor de "total_income", também está, o que pode indicar que os valores não foram informados ou não puderam ser extraídos, também podendo ser dados correlacionados e estão simétricos.



- Atráves da conta de divisão entre os valores ausentes e os valores total feita no programa abaixo, foi possivel observar que é uma parcela que podem fazer diferença em um análise final.

- Através do programa acima foi possivel observar que a ausencia da coluna days_employed (tempo que o cliente trabalhou) está diretamente ligada a ausencia da coluna total_income (renda mensal), mas particularmente não consegui achar outra analogia de caracteristicas em outras colunas que possam ser relacionadas com a ausencia da coluna do tempo que o cliente trabalhou. 

- Atualmente será necessário retirar os valores ausentes para conseguir fazer uma análise mais completa da tabela e verificar o porque seus dados estão ausentes se foi de forma proposital ou não.

In [14]:
#Calcular a porcentagem dos valores ausentes em comparação ao conjunto de dados 

porcentagem =(dados.isna().sum() / dados.shape[0]) *100

print(porcentagem)

children             0.000000
days_employed       10.099884
dob_years            0.000000
education            0.000000
education_id         0.000000
family_status        0.000000
family_status_id     0.000000
gender               0.000000
income_type          0.000000
debt                 0.000000
total_income        10.099884
purpose              0.000000
dtype: float64


- Foi possivel verificar que a porcentagem de valores ausentes em comparação com os dados são de 10%, um valor relativamente alto para numeros ausentes que podem complicar a analise de dados.

In [15]:
# Vamos investigar clientes que não possuem dados sobre as características identificadas e a coluna com os valores ausentes

#Investigando se os dados apresentam informações relevantes quanto à children
print(dados[days_employed_null]['children'].value_counts().sort_values())

print()

#Investigando se os dados apresentam informações relevantes quanto à education
print(dados[days_employed_null]['education'].value_counts().sort_values())

print()

#Investigando se os dados apresentam informações relevantes quanto à dob_years
print(dados[days_employed_null]['dob_years'].value_counts().sort_values())

print()

#Investigando se os dados apresentam informações relevantes quanto à family_status
print(dados[days_employed_null]['family_status'].value_counts().sort_values())

print()

#Investigando se os dados apresentam informações relevantes quanto à gender
print(dados[days_employed_null]['gender'].value_counts().sort_values())

print()

#Investigando se os dados apresentam informações relevantes quanto à debt
print(dados[days_employed_null]['debt'].value_counts().sort_values())

 5        1
-1        3
 4        7
 20       9
 3       36
 2      204
 1      475
 0     1439
Name: children, dtype: int64

Primary Education         1
PRIMARY EDUCATION         1
SOME COLLEGE              7
Some College              7
primary education        19
BACHELOR'S DEGREE        23
Bachelor's Degree        25
some college             55
Secondary Education      65
SECONDARY EDUCATION      67
bachelor's degree       496
secondary education    1408
Name: education, dtype: int64

73     1
19     1
72     2
70     3
71     5
20     5
69     5
68     9
0     10
67    16
22    17
21    18
65    20
66    20
24    21
25    23
63    29
59    34
26    35
23    36
27    36
32    37
64    37
62    38
61    38
60    39
44    44
53    44
48    46
46    48
55    48
43    50
29    50
49    50
45    50
51    50
50    51
33    51
39    51
52    53
37    53
56    54
38    54
54    55
57    56
58    56
28    57
30    58
41    59
47    59
36    63
35    64
42    65
31    65
40    66
34    69
Nam

In [16]:
# Verificar a distribuição

# Verificar a distribuição de quantidade de crianças que os clientes tem

print(dados['children'].value_counts().sort_values())

print()

# Verificar a distribuição de quantidade de tempo que o cliente trabalhou
print(dados['days_employed'].value_counts().sort_values())

print()

# Verificar a distribuição de quantidade das idades dos clientes
print(dados['dob_years'].value_counts().sort_values())

print()

# Verificar a distribuição de quantidade de tipos de educação dos clientes
print(dados['education'].value_counts().sort_values())

print()

# Verificar a distribuição de quantidade de tipos de identificador da educação do cliente 
print(dados['education_id'].value_counts().sort_values())

print()

# Verificar a distribuição do estado civil do cliente
print(dados['family_status'].value_counts().sort_values())

print()

# Verificar a distribuição de  identificador do estado civil do cliente
print(dados['family_status_id'].value_counts().sort_values())

print()
# Verificar a distribuição dos generos  dos clientes
print(dados['gender'].value_counts().sort_values())

print()

# Verificar a distribuição de se o cliente já deixou de pagar um empréstimo
print(dados['debt'].value_counts().sort_values())

print()

# Verificar a distribuição de o tipo de renda do cliente
print(dados['income_type'].value_counts().sort_values())

print()
# Verificar a distribuição de renda mensal do cliente
print(dados['total_income'].value_counts().sort_values())

print()

# Verificar a distribuição de motivo para fazer um empréstimo
print(dados['purpose'].value_counts().sort_values())

 5         9
 4        41
-1        47
 20       76
 3       330
 2      2055
 1      4818
 0     14149
Name: children, dtype: int64

-327.685916       1
-1333.589790      1
-147.097779       1
-523.142699       1
-4664.837665      1
                 ..
-4994.763109      1
 346393.178654    1
-935.373324       1
-2916.709461      1
-3382.113891      1
Name: days_employed, Length: 19351, dtype: int64

75      1
74      6
73      8
19     14
72     33
20     51
71     58
70     65
69     85
68     99
0     101
21    111
67    167
22    183
66    183
65    194
23    254
24    264
64    265
63    269
62    352
61    355
25    357
60    377
26    408
55    443
59    444
51    448
53    459
57    460
58    461
46    475
54    479
47    480
52    484
56    487
27    493
45    497
28    503
49    508
32    510
43    513
50    514
37    537
48    538
30    540
29    545
44    547
36    555
31    560
39    573
33    581
42    597
38    598
34    603
41    607
40    609
35    617
Name: dob_years,

**Descrevendo minhas descobertas.**

- Comparando a distribuição com os valores de cada coluna na base principal,pode se notar que os valores ausentes estão relacionados em sua grande parte com pessoas casadas, sem filho, mulheres e com o debito em dia.Porém é notavel que em relação a outros dados se torna aleatório sem nenhum padrão.

- Os valores ausentes apenas tem relação entre as duas colunas com os valores o qual são a 'days_employed' e a 'total_income'

**Possíveis motivos para valores ausentes nos dados**

- O motivo pode ser ja que como não tem o tempo que o cliente trabalhou então não tem uma renda mensal pois se o cliente não trabalhou como diz o dado da tabela , não é possivel que ele tenha uma renda mensal. porem , não existe padrão com as outras colunas.Os valores provavelmente estão ausentes de uma forma aleatória sem nenhum padrão.


**começar a verificar se os valores ausentes são aleatórios.**

- Não foi encontrado nenhum padrão após análise, pois os valores ausentes não parecem estar diretamente ligados a nenhum outro valor de outras colunas com exceção da renda mensal.

**Conclusões**

- Não foi encontrado nenhum padrão após análise, pois os valores ausentes não parecem estar diretamente ligados a nenhum outro valor de outras colunas

- Em relação aos valores ausentes atraves do método fillna() serão retirados

- Atraves de alguns métodos serão percebidos se existe linhas duplicadas, valores escritos de modo diferente mas signifcando a mesma situação  entre outros metodos para identificar erros na tabela e deixar ela limpa para o processo final.


## Transformação de dados


In [17]:
# Vamos ver todos os valores na coluna de educação para verificar se e quais grafias precisarão ser corrigidas

print(dados['education'].unique())


["bachelor's degree" 'secondary education' 'Secondary Education'
 'SECONDARY EDUCATION' "BACHELOR'S DEGREE" 'some college'
 'primary education' "Bachelor's Degree" 'SOME COLLEGE' 'Some College'
 'PRIMARY EDUCATION' 'Primary Education' 'Graduate Degree'
 'GRADUATE DEGREE' 'graduate degree']


In [18]:
# Corrija os registros, se necessário
#Foi possivel observar que alguns valores estão querendo dizer a mesma coisa porem como um esta com letra maisucula e outro minuscula o prohgrma entende que são situações diferentes

dados['education']=dados['education'].str.lower()

# Verificando todos os valores na coluna para ter certeza de que os corrigimos

print(dados['education'].unique())



["bachelor's degree" 'secondary education' 'some college'
 'primary education' 'graduate degree']


In [19]:
# Vamos ver a distribuição de valores na coluna `children`

soma_crianças= dados['children'].value_counts()

soma_crianças


 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

- Existe um problema , devido ao fato que existe linhas com o numero negativo, e não pode existir uma quantidade de criança com numero negativo, podem ter ococrrido por erro de digitação onde na verdade era para ser a quantidade de 1 criança.Alem de existir dados também de fámilias com 20 filhos

In [20]:
#Porcentagem de dados problemáticos na coluna chilndren)

porcentagem_criança  =  (soma_crianças/soma_crianças.sum()) *100
print(porcentagem_criança)

#temos uma porcentagem baixa de 0.21% de valores negativos e também baixa para valores com 20 filhos


 0     65.732869
 1     22.383275
 2      9.547038
 3      1.533101
 20     0.353078
-1      0.218351
 4      0.190476
 5      0.041812
Name: children, dtype: float64


In [21]:
# [corrija os dados com base na sua decisão]

# corrijindo os dados com base que o numero -1 seria 1
dados['children']=dados['children'].replace(-1,1)


# corrijindo os dados com base que o numero 20 seria 2
dados['children']=dados['children'].replace(20,2)


In [22]:
# Verificar a coluna `children` novamente para ter certeza de que está tudo corrigido

print(dados['children'].unique())

print()

print(dados['children'].value_counts())

[1 0 3 2 4 5]

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64


- Grande maioria dos dados estão negativos na coluna `days_employed` ,não podendo ter esse valor negativo é necessario transformar os valores negativos para positivos.

In [23]:
# Encontre dados problemáticos em `days_employed`, se existirem, e calcule a porcentagem

dados['days_employed'].describe()


count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

- É possivel verificar que existe numeros negativos , além que existem numeros com uma quantidade muito alta, podendo assim verificar que aconteceu algum erro de digitação ou apenas erros nos valores.

In [24]:
# Aborde os valores problemáticos, se existirem

dados['days_employed']=dados['days_employed'].abs()
    

In [25]:
# Verifique o resultado - certifique-se de que está corrigido
dados['days_employed'].describe()


count     19351.000000
mean      66914.728907
std      139030.880527
min          24.141633
25%         927.009265
50%        2194.220567
75%        5537.882441
max      401755.400475
Name: days_employed, dtype: float64

In [26]:
#verificar problemas na coluna 'days_employed'

print(dados['days_employed'].max())

print(dados['dob_years'].max())


401755.40047533
75


- A maior idade registrada no banco de dados foi de 75 anos sendo impossivel uma pessoa ter trabalhado 401755.40047533 dias


- Aos valores que estão fora da realidade , irei trocar pela mediana dos numeros

In [27]:
# Resolva os problemas na coluna 'days_employed'`, se existirem
print('Quantidade de dias que uma pessoa de 75 anos viveu:', dados['dob_years'].max()*365)
print('Quantidade média de dias que uma pessoa de 75 anos trabalhou:',(75-18)*365)

x= (75-18)*365
tipicos_days_employed = dados[dados['days_employed'] < x]

# dias típicos trabalhados
medianatipicos_days_employed=tipicos_days_employed['days_employed'].median()
mediatipicos_days_employed=tipicos_days_employed['days_employed'].mean()

print()
print(f'{medianatipicos_days_employed}: mediana de dias trabalhados típicos')
print(f'{mediatipicos_days_employed} :média de dias trabalhados típicos')

# dias atípicos trabalhados
atipicos_days_employed = dados[dados['days_employed'] > x]
mediana_atipicos_days_employed=atipicos_days_employed['days_employed'].median()
media_atipicos_days_employed=atipicos_days_employed['days_employed'].mean()

print()
print(f'{mediana_atipicos_days_employed}: mediana de dias trabalhados atípicos')
print(f'{media_atipicos_days_employed} :média de dias trabalhados atípicos')

Quantidade de dias que uma pessoa de 75 anos viveu: 27375
Quantidade média de dias que uma pessoa de 75 anos trabalhou: 20805

1630.0193809778218: mediana de dias trabalhados típicos
2353.0159319988766 :média de dias trabalhados típicos

365213.3062657312: mediana de dias trabalhados atípicos
365004.3099162686 :média de dias trabalhados atípicos


In [28]:
# Verifique o resultado - certifique-se de que está corrigido

dados.loc[dados['days_employed'] > x ,'days_employed'] = medianatipicos_days_employed

dados['days_employed'].describe()


count    19351.000000
mean      2224.303043
std       2107.307642
min         24.141633
25%        927.009265
50%       1630.019381
75%       2747.423625
max      18388.949901
Name: days_employed, dtype: float64

- Se reparar agora a tabela esta como dias máximo trabalhados 18388.949901 dias e sua média foi para 2224.303043 dias

In [29]:
# Vamos ver os valores da coluna

print(dados['family_status'].describe())
print()

print(dados['family_status'].value_counts())

print()
print(dados['family_status'].unique())


count       21525
unique          5
top       married
freq        12380
Name: family_status, dtype: object

married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

['married' 'civil partnership' 'widow / widower' 'divorced' 'unmarried']


- Não foi verificado erro na coluna  'family_status'

In [30]:
# Vamos ver os valores na coluna

#veriifcar linhas duplicadas

print(dados['gender'].describe())

print()

print(dados['gender'].unique())

print()

print(dados['gender'].value_counts())

count     21525
unique        3
top           F
freq      14236
Name: gender, dtype: object

['F' 'M' 'XNA']

F      14236
M       7288
XNA        1
Name: gender, dtype: int64


In [31]:
# Aborde os valores problemáticos, se existirem
dados['gender'].isna().sum()

dados['gender']=dados['gender'].replace('XNA', 'sem genero')



 - Na coluna gênero, há "F", "M" e "XNA". "F" usualmente se refere à "Female" (Mulher), "M" se refere à "Male" (Homem), já "XNA" pode significar um terceiro gênero ou informação incorreta nos dados. Vamos analisar a linha completa.

In [32]:
dados[dados['gender'] == 'sem genero']

children  days_employed  dob_years     education  education_id  \
10701         0    2358.600502         24  some college             2   

           family_status  family_status_id      gender income_type  debt  \
10701  civil partnership                 1  sem genero    business     0   

       total_income          purpose  
10701     32624.825  buy real estate

- Não existe nehuma relação , troquei para sem genero para deixar a tabela mais explicativa.

In [33]:
# Verifique o resultado - certifique-se de que está corrigido
print(dados['gender'].unique())


['F' 'M' 'sem genero']


In [34]:
# Vamos ver os valores na coluna

print(dados['income_type'].describe())

print()

print(dados['income_type'].unique())

print()

print(dados['income_type'].value_counts())

count        21525
unique           8
top       employee
freq         11119
Name: income_type, dtype: object

['employee' 'retiree' 'business' 'civil servant' 'unemployed'
 'entrepreneur' 'student' 'paternity / maternity leave']

employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
unemployed                         2
entrepreneur                       2
paternity / maternity leave        1
student                            1
Name: income_type, dtype: int64


- Não foram localizados valores problematicos , se quiser poderia juntar business com entrepreneur mas como é um valor muito baixo não faria diferença para a nossa analise de dados


**Verificar duplicatas nos dados e como resolver.**

In [35]:
# Verificar duplicatas
print(dados.duplicated().sum())

71


Temos 71 linhas duplicadas

In [36]:
# Aborde as duplicatas, se existirem
dados=dados.drop_duplicates().reset_index(drop= True)

In [37]:
# Última verificação se temos duplicatas
dados.duplicated().sum()

0

In [38]:
#Verifique o tamanho do conjunto de dados que você tem agora após suas primeiras manipulações com ele

print(f'{dados.shape[0]} linhas e {dados.shape[1]} colunas')

21454 linhas e 12 colunas


- Antes a tabela possuia 21525 e atualmente possui 21454 linhas , dividindo o total anterior pelo atual é possivel notar que temos uma diferença de 0.4% de dimuição de linhas da tabela devido a retirada das linhas duplicadas


In [39]:
# verificando informações da coluna 'dob_years'

print(sorted(dados['dob_years'].unique()))

print()

print((dados['dob_years'].value_counts().sort_values()))
    


[0, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75]

75      1
74      6
73      8
19     14
72     33
20     51
71     56
70     65
69     85
68     99
0     101
21    111
67    167
66    182
22    183
65    193
23    252
64    260
24    264
63    269
62    348
61    354
25    357
60    374
26    408
55    443
59    443
51    446
58    454
57    456
53    459
46    472
54    476
47    477
56    483
52    484
27    493
45    496
28    503
49    508
32    509
43    512
50    513
37    536
48    536
30    537
29    544
44    545
36    554
31    559
39    572
33    581
42    596
38    597
34    601
41    605
40    607
35    616
Name: dob_years, dtype: int64


In [40]:
#conte a porcentagem dos valors suspeitos
porcentagem_dob_years = dados.loc[dados['dob_years'] == 0 ]['dob_years'].count().sum()/dados['dob_years'].value_counts().sum()

print(f'A porcentagem é de {porcentagem_dob_years:.2%}')

A porcentagem é de 0.47%


In [41]:
#transformação da idade 0 na média da coluna
# Resolva os problemas na coluna `dob_years`, se existirem

mean_age= dados[dados['dob_years'] > 0]['dob_years'].mean()
median_age= dados[dados['dob_years'] > 0]['dob_years'].median()

print( f' A média é {mean_age} e a mediana é {median_age}')

print()

dados['dob_years']=dados['dob_years'].replace(0,43)

 A média é 43.47590502505503 e a mediana é 43.0



In [42]:
print(dados['dob_years'].unique())

[42 36 33 32 53 27 43 50 35 41 40 65 54 56 26 48 24 21 57 67 28 63 62 47
 34 68 25 31 30 20 49 37 45 61 64 44 52 46 23 38 39 51 59 29 60 55 58 71
 22 73 66 69 19 72 70 74 75]


## Trabalhando com valores ausentes

In [43]:
# Encontre os dicionários
dados.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

In [44]:
educação_dicionários= dados[['education','education_id']]
educação_dicionários= educação_dicionários.drop_duplicates().reset_index(drop= True)
educação_dicionários

education  education_id
0    bachelor's degree             0
1  secondary education             1
2         some college             2
3    primary education             3
4      graduate degree             4

In [45]:

familia_dicionario=dados[['family_status', 'family_status_id']]
familia_dicionario= familia_dicionario.drop_duplicates().reset_index(drop= True)
familia_dicionario

family_status  family_status_id
0            married                 0
1  civil partnership                 1
2    widow / widower                 2
3           divorced                 3
4          unmarried                 4

### Restaurar valores ausentes em `total_income`

- Criando a categoria de idade para os clientes e assim criando uma nova coluna de categoria de idade

In [46]:
# Vamos escrever uma função que calcule a categoria de idade
def calcular_idade(caracter_idade):
    idade=int(caracter_idade)
    
    if idade <= 30 :
        return 'early adulthood'
    if idade > 30  and idade <=40:
        return 'middle adulthood'
    if idade > 40 and idade <=50 :
        return 'senior adulthood'
    if idade > 50 and idade <=60:
        return 'young old'
    if idade > 60 and idade <=70:
        return 'median old' 
    if idade  > 70 :
        return 'old old'





In [47]:
# Teste se a função funciona
print(calcular_idade( caracter_idade = '20' ))
print(calcular_idade( caracter_idade = '30' ))
print(calcular_idade( caracter_idade = '40' ))
print(calcular_idade( caracter_idade = '50' ))
print(calcular_idade( caracter_idade = '60' ))
print(calcular_idade( caracter_idade = '70' ))
print(calcular_idade( caracter_idade = '80' ))




early adulthood
early adulthood
middle adulthood
senior adulthood
young old
median old
old old


In [48]:
# Criar coluna nova com base na função 

dados['group_years']= dados['dob_years'].apply(calcular_idade)

In [49]:
# Verificar como os valores na nova coluna
display(dados.head(10))


children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0    1630.019381         53  secondary education             1   
5         0     926.185831         27    bachelor's degree             0   
6         0    2879.202052         43    bachelor's degree             0   
7         0     152.779569         50  secondary education             1   
8         2    6929.865299         35    bachelor's degree             0   
9         0    2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose       group_years  
0                purchase of the house  senior adulthood  
1                         car purchase  middle adulthood  
2                purchase of the house  middle adulthood  
3              supplementary education  middle adulthood  
4                    to have a wedding         young old  
5                purchase of the house   early adulthood  
6                 housing transactions  senior adulthood  
7                            education  senior adulthood  
8                     having a wedding  middle adulthood  
9  purchase of the house for my family  senior adulthood

- Criando uma tabela com dados sem valores ausentes onde apenas as colunas 'days_employed' e  'total_income' possuem

In [50]:
# Crie uma tabela sem valores ausentes e exiba algumas de suas linhas para garantir que ela fique boa
dados_sem_ausentes= dados.dropna()
dados_sem_ausentes.isna().sum()


children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
group_years         0
dtype: int64

In [51]:
# Veja os valores medianos e média  de renda com base em seus fatores identificados

display(dados_sem_ausentes.groupby(['group_years'])['total_income'].median().sort_values())
dados_sem_ausentes.groupby(['group_years'])['total_income'].mean().sort_values()



group_years
old old             18611.5935
median old          19705.8550
young old           22056.7710
early adulthood     22957.1850
senior adulthood    24563.6500
middle adulthood    24825.1865
Name: total_income, dtype: float64

group_years
old old             19575.454327
median old          23245.390243
young old           25482.856294
early adulthood     25817.674826
senior adulthood    28332.806009
middle adulthood    28376.735148
Name: total_income, dtype: float64

- Pode se notar que a maior média/mediana da renda mensal são das pessoas entre 30 e 40 anos e as menores são das pessoas com mais de 70 anos

In [52]:
# Veja os valores medianos e media days_employed com base em seus fatores identificados

display(dados_sem_ausentes.groupby(['group_years'])['days_employed'].median().sort_values())
dados_sem_ausentes.groupby(['group_years'])['days_employed'].mean().sort_values()

group_years
early adulthood     1046.095163
median old          1630.019381
middle adulthood    1630.019381
old old             1630.019381
young old           1630.019381
senior adulthood    1931.819208
Name: days_employed, dtype: float64

group_years
early adulthood     1280.411387
old old             1912.126700
middle adulthood    2087.434418
median old          2102.589331
young old           2613.158806
senior adulthood    2752.828991
Name: days_employed, dtype: float64

- Pode se notar que a maior média/mediana de dias trabalhados são das pessoas entre 40 e 50 anos e as menores são das pessoas com menos de 30 anos.

In [53]:
# Veja os valores medianos e média dob_years com base em seus fatores identificados
display(dados_sem_ausentes.groupby(['group_years'])['dob_years'].median().sort_values())
dados_sem_ausentes.groupby(['group_years'])['dob_years'].mean().sort_values()

group_years
early adulthood     27.0
middle adulthood    35.5
senior adulthood    45.0
young old           55.0
median old          64.0
old old             71.0
Name: dob_years, dtype: float64

group_years
early adulthood     26.532383
middle adulthood    35.561315
senior adulthood    45.308153
young old           55.406366
median old          64.146234
old old             71.683673
Name: dob_years, dtype: float64

In [54]:
# Verificando os valores medianos de renda com base no fator tipo de renda
display(dados_sem_ausentes.groupby(['income_type'])['total_income'].median().sort_values())

# Verificando os valores médios de renda com base no fator tipo de renda
display(dados_sem_ausentes.groupby(['income_type'])['total_income'].mean().sort_values())

income_type
paternity / maternity leave     8612.6610
student                        15712.2600
retiree                        18962.3180
unemployed                     21014.3605
employee                       22815.1035
civil servant                  24071.6695
business                       27577.2720
entrepreneur                   79866.1030
Name: total_income, dtype: float64

income_type
paternity / maternity leave     8612.661000
student                        15712.260000
unemployed                     21014.360500
retiree                        21940.394503
employee                       25820.841683
civil servant                  27343.729582
business                       32386.793835
entrepreneur                   79866.103000
Name: total_income, dtype: float64

- Pode se notar que a maior média/mediana de renda mensal são das pessoas que trabalham com negócios e são empresários  e as menores são das pessoas com licença maternidade/paternidade.

In [55]:
#veja os valores médios de renda com base em seus fatores identificados
dados_sem_ausentes.pivot_table(index='family_status',columns='group_years',values='total_income',aggfunc='mean')

group_years        early adulthood    median old  middle adulthood  \
family_status                                                        
civil partnership     25787.471809  22915.379770      27961.994630   
divorced              26426.119818  26427.673658      28973.893349   
married               25719.823516  22990.098107      28641.701590   
unmarried             25946.194951  23642.547661      27501.397397   
widow / widower       25030.870857  22971.860510      27457.455415   

group_years             old old  senior adulthood     young old  
family_status                                                    
civil partnership  18718.380929      28694.861508  24616.127094  
divorced           22036.970500      27996.732159  25198.240874  
married            19922.862644      28255.457394  26051.833940  
unmarried          21317.738200      29595.352542  26806.520576  
widow / widower    18339.324320      24291.203075  22214.677160

In [56]:
#veja os valores medianos de renda com base em seus fatores identificados

dados_pivot=dados_sem_ausentes.pivot_table(index=['income_type','group_years'],columns='family_status',values='total_income',aggfunc='median')
display(dados_pivot)

family_status                                 civil partnership    divorced  \
income_type                 group_years                                       
business                    early adulthood          26121.1775  27060.1570   
                            median old               24419.3790  30949.2265   
                            middle adulthood         26321.3385  30785.7320   
                            old old                  24259.6870  26948.7450   
                            senior adulthood         29807.5000  26621.8525   
                            young old                26068.2560  28146.4810   
civil servant               early adulthood          22744.6060  23948.3365   
                            median old               20030.5470  31418.9035   
                            middle adulthood         25169.6800  24644.5360   
                            old old                         NaN  14536.1310   
                            senior adulthood         25556.9480  28147.2310   
                            young old                24482.7440  21093.3560   
employee                    early adulthood          20995.6060  20109.8000   
                            median old               22787.9430  25569.6020   
                            middle adulthood         23338.4400  22842.5350   
                            old old                  15831.3450         NaN   
                            senior adulthood         23308.9575  24237.5525   
                            young old                22880.8430  20861.7140   
entrepreneur                early adulthood          79866.1030         NaN   
paternity / maternity leave middle adulthood                NaN         NaN   
retiree                     early adulthood          12807.0710  12066.7990   
                            median old               19386.9750  18693.7630   
                            middle adulthood         18201.6450  18454.4280   
                            old old                  16694.3650  23331.5030   
                            senior adulthood         24160.4520  32975.3170   
                            young old                18122.3020  20185.3180   
student                     early adulthood                 NaN         NaN   
unemployed                  middle adulthood                NaN         NaN   
                            senior adulthood         32435.6020         NaN   

family_status                                    married   unmarried  \
income_type                 group_years                                
business                    early adulthood   25641.8430  25348.7140   
                            median old        28442.1420  29993.0000   
                            middle adulthood  29025.0380  29159.2735   
                            old old           29329.0450  42271.4370   
                            senior adulthood  28272.7130  29871.0470   
                            young old         28469.4910  26786.9020   
civil servant               early adulthood   22869.1900  24617.5440   
                            median old        26833.1360  35849.2295   
                            middle adulthood  24663.0880  26838.5520   
                            old old                  NaN         NaN   
                            senior adulthood  22971.2210  26479.9450   
                            young old         24220.7485  21020.2380   
employee                    early adulthood   22023.2830  22090.9810   
                            median old        26548.8095  19578.2860   
                            middle adulthood  23572.5450  21986.8040   
                            old old           26394.2450         NaN   
                            senior adulthood  22855.3725  23511.2070   
                            young old         23123.2985  21627.1650   
entrepreneur                early adulthood          NaN         NaN   
paternity / maternity leave middle adu

- Se os dados fossem muito parecidos seria melhor usar a média mas como existe numeros diversos e variados a mediana de fato será a melhor opção para preencher os valores ausentes.

In [57]:
def fill_na_total_income (dados, col, parâmetro):
    dados_aux = dados.groupby([parâmetro])[col].median().reset_index()
    
    for row in range(dados_aux.shape[0]):
        grupo = dados_aux.loc[row, parâmetro] 
        valor = dados_aux.loc[row, col] 
        
        #Preenchendo os valores vazios do grupo com o valor da mediana para esse grupo
        dados.loc[dados[parâmetro]==grupo, col] = dados.loc[dados[parâmetro]==grupo, col].fillna(valor)
    
    return dados 

In [58]:
#Usando a função
fill_na_total_income (dados,'total_income', 'gender')
fill_na_total_income (dados,'total_income', 'education')
fill_na_total_income (dados,'total_income', 'group_years')

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0    1630.019381         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21449         1    4529.316663         43  secondary education             1   
21450         0    1630.019381         67  secondary education             1   
21451         1    2113.346888         38  secondary education             1   
21452         3    3112.481705         38  secondary education             1   
21453         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21449  civil partnership                 1      F    business     0   
21450            married                 0      F     retiree     0   
21451  civil partnership                 1      M    employee     1   
21452            married                 0      M    employee     1   
21453            married                 0      F    employee     0   

       total_income                  purpose       group_years  
0         40620.102    purchase of the house  senior adulthood  
1         17932.802             car purchase  middle adulthood  
2         23341.752    purchase of the house  middle adulthood  
3         42820.568  supplementary education  middle adulthood  
4         25378.572        to have a wedding         young old  
...             ...                      ...               ...  
21449     35966.698     housing transactions  senior adulthood  
21450     24959.969        purchase of a car        median old  
21451     14347.610                 property  middle adulthood  
21452     39054.888        buying my own car  middle adulthood  
21453     13127.587             to buy a car  middle adulthood  

[21454 rows x 13 columns]

In [59]:
# Verificando se temos algum erro
print(dados['total_income'].isna().sum())

dados['total_income'].describe()

0


count     21454.000000
mean      26436.254298
std       15703.063756
min        3306.762000
25%       17219.817250
50%       22710.927000
75%       31330.237250
max      362496.645000
Name: total_income, dtype: float64

- verificando se o número total de valores nesta coluna corresponde ao número de valores em outras.

In [60]:
# Verificar o número de entradas nas colunas
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21454 non-null  int64  
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      21454 non-null  float64
 11  purpose           21454 non-null  object 
 12  group_years       21454 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


como 21454 - 19351 = 2103 é o numero de valores ausentes
O número total de valores da coluna total_income corresponde ao número de valores das outras colunas, não restando valores nulos.

###  Restaurar valores em `days_employed`

In [61]:
# Distribuição de `days_employed` medianos com base em seus parâmetros identificados

print(dados_sem_ausentes.groupby('group_years')['days_employed'].median())


group_years
early adulthood     1046.095163
median old          1630.019381
middle adulthood    1630.019381
old old             1630.019381
senior adulthood    1931.819208
young old           1630.019381
Name: days_employed, dtype: float64


In [62]:
# Distribuição de `days_employed` médios com base em seus parâmetros identificados

print(dados_sem_ausentes.groupby('group_years')['days_employed'].mean())


group_years
early adulthood     1280.411387
median old          2102.589331
middle adulthood    2087.434418
old old             1912.126700
senior adulthood    2752.828991
young old           2613.158806
Name: days_employed, dtype: float64


- Os valores de médias e mediandas estão muito diferentes, o que indica valores atípicos nos dados. Baseado nisso, vamos utilizar a mediana.

In [63]:
# Vamos escrever uma função que calcule  medianas  com base no seu parâmetro identificado
def fill_na_days_employed (dados, col, parâmetro):
    dados_aux = dados.groupby([parâmetro])[col].median().reset_index()
    
    for row in range(dados_aux.shape[0]):
        grupo = dados_aux.loc[row, parâmetro] 
        valor = dados_aux.loc[row, col] 
        
        #Preenchendo os valores vazios do grupo com o valor da mediana para esse grupo
        dados.loc[dados[parâmetro]==grupo, col] = dados.loc[dados[parâmetro]==grupo, col].fillna(valor)
    
    return dados

In [64]:
# Verifique se a função funciona
fill_na_days_employed (dados,'days_employed', 'group_years')


children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0    1630.019381         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21449         1    4529.316663         43  secondary education             1   
21450         0    1630.019381         67  secondary education             1   
21451         1    2113.346888         38  secondary education             1   
21452         3    3112.481705         38  secondary education             1   
21453         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21449  civil partnership                 1      F    business     0   
21450            married                 0      F     retiree     0   
21451  civil partnership                 1      M    employee     1   
21452            married                 0      M    employee     1   
21453            married                 0      F    employee     0   

       total_income                  purpose       group_years  
0         40620.102    purchase of the house  senior adulthood  
1         17932.802             car purchase  middle adulthood  
2         23341.752    purchase of the house  middle adulthood  
3         42820.568  supplementary education  middle adulthood  
4         25378.572        to have a wedding         young old  
...             ...                      ...               ...  
21449     35966.698     housing transactions  senior adulthood  
21450     24959.969        purchase of a car        median old  
21451     14347.610                 property  middle adulthood  
21452     39054.888        buying my own car  middle adulthood  
21453     13127.587             to buy a car  middle adulthood  

[21454 rows x 13 columns]

In [65]:
# Substituir valores ausentes
# Verificando se a função funcionou
print(dados['days_employed'].isna().sum())

dados['days_employed'].describe()


0


count    21454.000000
mean      2163.754556
std       2011.685499
min         24.141633
25%       1023.709707
50%       1630.019381
75%       2523.810846
max      18388.949901
Name: days_employed, dtype: float64

In [66]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     21454 non-null  float64
 2   dob_years         21454 non-null  int64  
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      21454 non-null  float64
 11  purpose           21454 non-null  object 
 12  group_years       21454 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


## Categorização de dados

- Atraves de funções foi feito uma nova coluna para propósito devido ao fato que varias colunas estavam querendo dizer o mesmo propósito só que escrito de maneira diferente.

- Foi feito também uma nova coluna para a renda mensal caracterizando o nivel que cada pessoa esta de acordo com sua renda para assim poder fazer uma melhor analise dos dados.

In [67]:
# Verificando os valores para motivação para empréstimo

dados['purpose'].value_counts()

wedding ceremony                            791
having a wedding                            768
to have a wedding                           765
real estate transactions                    675
buy commercial real estate                  661
housing transactions                        652
buying property for renting out             651
transactions with commercial real estate    650
housing                                     646
purchase of the house                       646
purchase of the house for my family         638
construction of own property                635
property                                    633
transactions with my real estate            627
building a real estate                      624
buy real estate                             621
purchase of my own house                    620
building a property                         619
housing renovation                          607
buy residential real estate                 606
buying my own car                       

In [68]:
# Verificando os valores exclusivos para motivação para empréstimo
dados['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

Com base nos dados plotados é possivel verificar alguns valores que estão escritos diferentes mas querem dizer a mesma coisa
- Fazer um casamento
- Comprar uma casa
- comprar um carro
- Educação


In [69]:
# Vamos escrever uma função para categorizar os dados com base em tópicos comunss
def purpose_category(row):
    if 'hous' in row['purpose'] or 'property' in row['purpose'] or  'real' in row['purpose'] :
        return 'house'
    if 'wedd'in row['purpose']:
        return 'wedding'
    if 'car' in row['purpose']:
        return 'car'
    if 'education' in row['purpose']:
        return 'education'
print(purpose_category(dados.loc[2]))
print(purpose_category(dados.loc[3]))
print(purpose_category(dados.loc[4]))
print(purpose_category(dados.loc[50]))


house
education
wedding
car


In [70]:
# Crie uma coluna com as categorias e conte os valores para elas

dados['purpose_group']= dados.apply(purpose_category,axis=1)

dados.head(10)




children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0    1630.019381         53  secondary education             1   
5         0     926.185831         27    bachelor's degree             0   
6         0    2879.202052         43    bachelor's degree             0   
7         0     152.779569         50  secondary education             1   
8         2    6929.865299         35    bachelor's degree             0   
9         0    2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose       group_years purpose_group  
0                purchase of the house  senior adulthood         house  
1                         car purchase  middle adulthood           car  
2                purchase of the house  middle adulthood         house  
3              supplementary education  middle adulthood     education  
4                    to have a wedding         young old       wedding  
5                purchase of the house   early adulthood         house  
6                 housing transactions  senior adulthood         house  
7                            education  senior adulthood     education  
8                     having a wedding  middle adulthood       wedding  
9  purchase of the house for my family  senior adulthood         house

In [71]:
# Examinar todos os dados numéricos em sua coluna selecionada para categorização
print(dados['purpose_group'].value_counts())


house        10811
car           4306
education     3109
wedding       2324
Name: purpose_group, dtype: int64


In [72]:
# Obter estatísticas resumidas para a coluna
dados['purpose_group'].describe()

count     20550
unique        4
top       house
freq      10811
Name: purpose_group, dtype: object

**Categorizando dados de renda**

In [73]:
# Analisando a a coluna renda
dados['total_income'].describe()

count     21454.000000
mean      26436.254298
std       15703.063756
min        3306.762000
25%       17219.817250
50%       22710.927000
75%       31330.237250
max      362496.645000
Name: total_income, dtype: float64

In [74]:
# Criar função para categorização em diferentes grupos numéricos com base em intervalos
def tipos_renda(income):
     if  income <= 5000.0000000:
        return 'level 1'
     if  income <= 20000.000000:
        return 'level 2'
     if  income <= 30000.000000:
        return 'level 3'
     if  income <= 50000.000000:
        return 'level 4'
     if  income <= 100000.645000:
        return 'level 5' 
     if  income <= 362496.645000:
        return 'level 6'
    
print(tipos_renda( income = 20342))    
print(tipos_renda( income = 3400))
print(tipos_renda( income = 22000 ))
print(tipos_renda( income = 25000))
print(tipos_renda( income = 40000 ))
print(tipos_renda( income = 100000))
print(tipos_renda( income = 300000))

       

level 3
level 1
level 3
level 3
level 4
level 5
level 6


- Para dividir os criterios em níveis me basiei no metodo describe , para assim descobrir o numero mínimo, médio e maximo das rendas mensal dos clientes e conseguir dividir de acordo com esses valores , como o numero mínimo de renda era 3306.762000, coloquei no nivel 1 , como o nivel médio de renda era de 22710.927000 coloquei entre o nivel 3 e 4 , e o numero máximo 362496.645000, inserie entre o nivel 5 e 6

In [75]:
# Criar coluna com categorias
dados['income_group']= dados['total_income'].apply(tipos_renda)

dados.head(10)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0    1630.019381         53  secondary education             1   
5         0     926.185831         27    bachelor's degree             0   
6         0    2879.202052         43    bachelor's degree             0   
7         0     152.779569         50  secondary education             1   
8         2    6929.865299         35    bachelor's degree             0   
9         0    2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose       group_years purpose_group  \
0                purchase of the house  senior adulthood         house   
1                         car purchase  middle adulthood           car   
2                purchase of the house  middle adulthood         house   
3              supplementary education  middle adulthood     education   
4                    to have a wedding         young old       wedding   
5                purchase of the house   early adulthood         house   
6                 housing transactions  senior adulthood         house   
7                            education  senior adulthood     education   
8                     having a wedding  middle adulthood       wedding   
9  purchase of the house for my family  senior adulthood         house   

  income_group  
0      level 4  
1      level 2  
2      level 3  
3      level 4  
4      level 3  
5      level 4  
6      level 4  
7      level 3  
8      level 2  
9      level 3

In [76]:
# Conte os valores de cada categoria para ver a distribuição

dados['income_group'].value_counts()

level 3    8166
level 2    7343
level 4    4599
level 5    1221
level 6      99
level 1      26
Name: income_group, dtype: int64

## Verificar as Hipóteses


**Existe uma correlação entre o numero de  filhos e do pagamento em dia?**

In [77]:
# Verifique os dados das crianças e do pagamento em dia
pivot_table_children= dados.pivot_table( index='children', columns= 'debt', values= 'gender', aggfunc= 'count')
print(pivot_table_children)

print()
# Calcular a taxa de inadimplência com base no número de filhos
taxa_de_inadimplência= pivot_table_children[1]/(pivot_table_children[0]+pivot_table_children[1]) *100
print(taxa_de_inadimplência.sort_values())

debt            0       1
children                 
0         13028.0  1063.0
1          4410.0   445.0
2          1926.0   202.0
3           303.0    27.0
4            37.0     4.0
5             9.0     NaN

children
0    7.543822
3    8.181818
1    9.165808
2    9.492481
4    9.756098
5         NaN
dtype: float64


**Conclusão**

- Todos os clientes com 5 filhos estão com suas dívidas quitadas
- Os clientes com 4 filhos são os que tem maior taxa de inadimplência
- Os clientes que não tem nenhum filho e 3 filhos tem a menor taxa de inadimplência

A relação pode se dizer que quem tem nenhum filho por não ter um gasto a mais no mês em sua grande maioria pagam suas dívidas, porem os que tem mais filho não tem nenehuma dívida isso pode ser devido ao fato que apenas 9 pessoas tem 5 filhos então os dados são poucos para poder ter uma análise acertiva.

**Existe uma correlação entre o status familiar e o pagamento em dia?**

In [78]:
# Verifique os dados de status da família e do pagamento em dia

pivot_table_family_status = dados.pivot_table(index='family_status' , columns = 'debt' , values = 'gender', aggfunc= 'count')

print(pivot_table_family_status)

print()
# Calcular a taxa padrão com base no status da família

taxa_de_inadimplência_family= (pivot_table_family_status[1]) / (pivot_table_family_status[0] + pivot_table_family_status[1])*100
print(taxa_de_inadimplência_family.sort_values())

debt                   0    1
family_status                
civil partnership   3763  388
divorced            1110   85
married            11408  931
unmarried           2536  274
widow / widower      896   63

family_status
widow / widower      6.569343
divorced             7.112971
married              7.545182
civil partnership    9.347145
unmarried            9.750890
dtype: float64


**Conclusão**

- A taxa de inadimplência é menor entre os clientes divorciados e viúvos.
- A taxa de inadimplência é média entre os clientes casados.
- A taxa de inadimplência é maior entre os clientes em união estável e não casados.


Pode se dizer que o solteiros tem maior taxa de inadiplencia devido ao fato de sua vida de aproveitar o momento e não ter um planejamento futuro devido ao fato de não ter um parceiro ou parceira 


**Existe uma correlação entre o status familiar e o pagamento em dia?**


In [79]:
# Verifique os dados do nível de renda e do pagamento em dia

pivot_table_total_income= dados.pivot_table( index= 'income_group' , columns = 'debt' , values= 'gender', aggfunc= 'count' )

print(pivot_table_total_income)

# Calcular a taxa de inadimplência com base no nível de renda

taxa_pivot_table_total_income =pivot_table_total_income[1]/(pivot_table_total_income [0]+ pivot_table_total_income [1]) *100

taxa_pivot_table_total_income.sort_values()


debt             0    1
income_group           
level 1         24    2
level 2       6737  606
level 3       7469  697
level 4       4255  344
level 5       1135   86
level 6         93    6


income_group
level 6    6.060606
level 5    7.043407
level 4    7.479887
level 1    7.692308
level 2    8.252758
level 3    8.535391
dtype: float64

**Conclusão**

- Ao analisar os dados é possivel perceber que as pessoas que estão no level 5 e 6 o qual  tem uma renda mensal maior tem a menor taxa de inadimplencia , e os clientes que possuem menor renda são os que tem maior taxa de inadimplencia.

**Como a finalidade do crédito afeta a taxa de inadimplência?**

In [75]:
pivot_table_purpose_group=dados.pivot_table(index='purpose_group' , columns= 'debt' , values='gender',aggfunc= 'count' )
pivot_table_purpose_group


debt               0    1
purpose_group            
car             3903  403
education       2821  288
house          10029  782
wedding         2138  186

In [76]:
# Confira os percentuais de inadimplência para cada finalidade de crédito e analise-os

taxa_purpose_group = pivot_table_purpose_group[1] / (pivot_table_purpose_group[0] + pivot_table_purpose_group[1]) *100

taxa_purpose_group.sort_values()

purpose_group
house        7.233373
wedding      8.003442
education    9.263429
car          9.359034
dtype: float64

**Conclusão**

- A taxa de inadimplência é menor entre os clientes que pretendem casar e comprar uma casa.
- A taxa de inadimplência é maior entre os clientes que pretendem se educar e comprar um carro.


# Conclusão Geral 

**Liste suas conclusões nesta seção final. Certifique-se de incluir todas as suas conclusões importantes que o levaram à maneira como você processou e analisou os dados. Cubra os valores ausentes, duplicatas e possíveis razões e soluções para artefatos problemáticos que você teve que resolver.**

- Primeiramente importamos a biblioteca pandas e a tabela do excel e asism abrimos a tabela no programa.
- Analisamos a quantidade de linhas e colunas.
- Analisamos as informações que a tabela poderia dar para nós através do método info().
- analisamos os valores ausentes da tabela e procurar os valores ausentes e verificar se estava comprometendo a analise dos dados.
- analisamos a porcentagem em relação a todos os dados dos valores ausentes.
- analisamos se existia alguma relação dos valores ausentes com as outras colunas.
- verificamos também palavras que significam a mesma coisa só que escrita de forma diferente.
- verificamos também palavras que significam a mesma coisa só que escrita de  letra maiúscula e minúscula.
- analisamos erros de digitação como o fato dos valores negativo tento assim que fazer a transformação
- analisamos valores que não faziam sentido como o fato das horas trabalhadas ser um valor impossível , além da quantidade de filhos como 20 e -1.
- Para os valores impossíveis foi substituto pela mediana o qual fazia mais sentido.
- verificamos também  as linhas duplicadas e fizemos uma nova contagem de linhas e colunas para verificar a diminuição de linhas.
- inserimos duas novas colunas para poder diferenciar ainda mais certas situações como o nível de renda e classificações das idades.
- foram feitos medias e medianas para quase todas as colunas e verificar seus valores.
- foi criada uma função para preencher os valores ausentes com as medianas.
- Por fim a caracterização de dados e a manipulações das tabelas para responder as hipóteses propostas através de das tabelas pivot onde conseguimos colocar em tabelas algumas variavés para tirar a conclusão final e entender as hipóteses propostas.


**Existe uma correlação entre o numero de filhos e o pagamento em dia?**

- Todos os clientes com 5 filhos estão com suas dívidas quitadas
- Os clientes com 4 filhos são os que tem maior taxa de inadimplência
- Os clientes que não tem nenhum filho e 3 filhos tem a menor taxa de inadimplência
- A relação pode se dizer que quem tem nenhum filho por não ter um gasto a mais no mês em sua grande maioria pagam suas dívidas sem problemas com outros gastos, porem os que tem mais filho não tem nenehuma dívida isso pode ser devido ao fato que apenas 9 pessoas tem 5 filhos então os dados são poucos para fazer uma análise mais completa .

**Existe uma correlação entre o status familiar e o pagamento em dia?**

- A taxa de inadimplência é menor entre os clientes divorciados e viúvos.
- A taxa de inadimplência é média entre os clientes casados.
- A taxa de inadimplência é maior entre os clientes em união estável e solteiros.
- Pode se dizer que o solteiros tem maior taxa de inadiplencia devido ao fato de sua vida de aproveitar o momento e não ter um planejamento futuro devido ao fato de não ter um parceiro ou parceira

**Existe uma relação entre o nível de renda e o pagamento em dia ?**

- Ao analisar os dados é possivel perceber que as pessoas que estão no level 5 e 6 o qual tem uma renda mensal maior tem a menor taxa de inadimplencia , e os clientes que possuem menor renda ( 1,2,3) são os que tem maior taxa de inadimplencia.